### ISSUE TITLE
Date/DateTime-Period arithmetic operators

### ISSUE THREAD

**Issue by quinnj** [quinnj/Datetime.jl#3]
Currently, (+,-) are the standard for the `Date` type, while (<<,>>) work for the `DateTime`. Using (+,-) with `DateTime` promotes the `DateTime` to a `Date` and performs the operation. The reasoning originally was that users should probably avoid doing `DateTime-Period` arithmetic operations as it can be potentially confusing/imprecise. Consider the following: 

``` julia
2008-02-29T00:00:00 + year(1)
```

Should it return `2009-02-28T00:00:00`, or `2009-03-01T00:00:00`, or `2009-02-27T23::59:59`? Basically, should `DateTime` arithmetic be done with "relative" periods, or "absolute" periods (i.e. 31536000 seconds for a year, no matter leap year, leap seconds etc.). Should we allow both with the use of different operators and avoid any auto-promotion to a `Date` type?

All things to consider.

**Comment by quinnj** [comment]
After mulling this over some more, I think the right way to go is the following:
- Use the (+,-) operators for `Date-Period` operations. Only `Year, Month, Day` Periods are allowed in these operations
- Use (+,-) operators for _calendarical arithmetic_ for `DateTime-Period` operations
- Use(<<,>>) operators for _timeline arithmetic_ for  `DateTime-Period` operations. For reference on the terminology, see [this page](http://noda-time.googlecode.com/hg/docs/userguide/arithmetic.html) that explains Noda-Time's approach to DateTime arithmetic. This is also similar to Joda-Time's concept of adding/subtracting _durations_ from DateTimes (timeline arithmetic) and adding/subtracting _periods_ from DateTimes (calendrical arithmetic).

I like this approach because in general, if programmers are operating on DateTimes, it's probably for shifting/lagging dates by a consistent period, rather than for doing subsequent duration calculations 
(i.e.

``` julia
newdate = olddate + years(1)
diff = newdate - olddate   
```

`diff` will be different values depending on presence of a leap year, leap seconds, daylight savings, etc.
In other words, don't be stupid and mix _calendrical arithmetic_ and _timeline arithmetic_! I think if we're very upfront with how these work and talk about potential pitfalls, it will go along way towards proper usability.

**Comment by quinnj** [comment]
[Here's](http://noda-time.blogspot.com/2010/11/joys-of-datetime-arithmetic.html) another excellent blog post talking about some of the complexities of Date/DateTime arithmetic.
